rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+
+
+We can experiment with a hybrid strategy that mirrors some variables and shards other variables. All the ops can still be mirrored on each replica. We can have a strategy like:
+
+```python
+class HybridStrategy(DistributionStrategy):
+  def __init__(strategy1, strategy2, strategy_chooser):
+    """A hybrid distribution strategy.
+
+    Args:
+      strategy1: the main default strategy object.
+      strategy2: another strategy object.
+      strategy_chooser: a callable that takes variable name, variable shape and
+        variable dtype and returns which strategy to apply for this variable.
+    """
+    pass
+```
+
+
+# Alternatives Considered
+
+
+### Embedding in Mirrored Architecture: Sharding or Mirroring?
+
+
+In the `MirroredStrategy` or `CollectiveAllReduceStrategy`, there are several ways of handling embeddings. However due to the fact that `dynamic_partition` and `dynamic_stitch` required by embedding lookups on sharded embeddings only have CPU kernels, we can not shard embeddings across GPUs.
+
+We'll discuss how embeddings will be handled in single host and multi-host cases.
+
+
+##### Single host
+
+
+On a single host, embeddings can be placed on host memory or mirrored on devices.
+
+When an embedding variable is small, it may benefit from being mirrored on devices. Lookup can be performed local to each device. Updates can be casted to dense tensors and use the existing allreduce primitive to exchange gradients. This can still be faster than running allgather on their corresponding sparse updates.
+
+When an embedding variable is larger, we'll need an efficient allgather primitive to exchange updates between devices. Alternatively, we can place it on host memory at the cost of transferring embeddings from host to devices in the forward pass and gathering updates from devices to host in the backward pass.
+
+When an embedding variable is even larger, placing it on host may be the only choice.
+
+**We'll start with mirroring them on devices.** In the future we will provide a mechanism to place embedding variables on host and partition it or ask users to use parameter server architecture if their embeddings are too large.
+
+
+##### Multiple hosts
+
+
+After excluding sharding across devices, there are still three ways of handling embeddings on multiple hosts:
+
+*   sharding across all workers' host memory;
+*   mirroring on all workers' host memory;
+*   mirroring on all replicas, i.e. devices.
+
+On multiple hosts, mirroring could be cheaper than sharding in terms of communication cost since it requires smaller number of communications although it transfers more data. This is true even compared to the optimal implementation of sharding. When updates to an embedding are not very sparse, converting them to dense updates and applying allreduce in the mirrored case can be faster than sharding which relies on all-to-all personalized communication or all-gather. Furthermore, mirroring is much easier to implement.
```

Comment:
Can we elaborate it more? I think this only holds true when the sparsity is not high.

Reply:
Maybe some sentence is confusing. I meant when some gradients are not very sparse, we can densify and all-reduce them, which would be faster than allgather. This is one of the benefits of mirroring embedding.